### ISSUE TITLE
Question about coalesce behavior with struct

### ISSUE THREAD

**Issue by MatthieuMEZIL** [dotnet/roslyn#12864]
I have the following struct:

```
public struct MyStruct
{
    public int? Value { get; private set; }

    public MyStruct(int? value)
    {
        this.Value = value;
    }

    public static implicit operator MyStruct(int? i)
    {
        return new MyStruct(2);
    }
}
```

Then when I call it with 
`var test = null ?? new MyStruct(1);`

test is of type MyStruct and test.Value equals 1. 
I would expect a compiler error because MyStruct cannot be null or, at very least, I would expect that test.Value would be equal to 2.

So is it a bug or am I missing something?

**Comment by TKharaishvili** [comment]
It seems to me that the compiler doesn't complain because `null` is a valid value for type `int?` and it perceives the expression as something similar to:

```C#
int? a = null;
var b = a ?? new MyStruct(1);
```

Which is valid because values of type `int?` are also the values of type `MyStruct`  because of your implicit operator definition.

As for why `test.Value` doesn't equal to 2, it's because of the semantics of the `??` operator. It chooses the right hand side of the expression `null ?? new MyStruct(1)` hence the implicit assignment method is never called.

**Comment by MatthieuMEZIL** [comment]
If the implicit conversion happens after the evaluation of ?? yes this behavior is fine.
But IMVHO I would expect the implicit conversion to happen before (as it is with arithmetic operators for example). Don't you?

**Comment by TKharaishvili** [comment]
I can't quite picture the scenario with arithmetic operators you mentioned.

The right hand side of the assignment gets computed first which is why the `??` is evaluated first. Then the assignment is performed which doesn't call into the conversion function because the value being assigned is already of type `MyStruct`.

Conversion can only happen at the moment of the assignment and the assignment is always the last in line.

You can even view it in the roslyn source where the operator precedence is defined:
http://source.roslyn.io/#Microsoft.CodeAnalysis.CSharp/Parser/LanguageParser.cs,043ee14a29f7c50b

It looks pretty logical to me...

**Comment by MatthieuMEZIL** [comment]
So you think that it is expected that with the following class:
```
public class C
{
    public static implicit operator C(int? i)
    {
        return null;
    }
}
```
`(int?)1 ?? new C()` returns null.

**Comment by TKharaishvili** [comment]
No. I find this example quite surprising, but not the first one.

**Comment by leppie** [comment]
It is definitely surprising that `(C)((int?)1) ?? new C()` behaves different to `(int?)1 ?? new C()`

**Comment by aelij** [comment]
According to the [spec](https://github.com/ljw1004/csharpspec/blob/gh-pages/expressions.md#the-null-coalescing-operator):

> The type of the expression `a ?? b` depends on which implicit conversions are available on the operands.

And:

> Otherwise, if `b` has a type `B` and an implicit conversion exists from `a` to `B`, the result type is `B`. At run-time, `a` is first evaluated. If `a` is not null, `a` is unwrapped to type `A0` (if `A` exists and is nullable) and converted to type `B`, and this becomes the result. Otherwise, `b` is evaluated and becomes the result.

(I recommend reading the whole section on the null coalescing operator.)

So, the expression `x ?? new MyStruct(1)` is translated to `x.HasValue ? (MyStruct)x : new MyStruct(1)`.

It seems that if `x` is the constant `null`, there's even an optimization that simply emits just the right expression (couldn't find a reference to that in the spec though).

This also explains the second example. It's translated to `((int?)1).HasValue ? (C)(int?)1 : new C()` (roughly; the `int?` conversion is saved to a temporary variable so it's only evaluated once).

Lastly, regarding `(C)((int?)1) ?? new C()` - the expression `(C)((int?)1)` requires no conversion, and evaluates to `null` at run-time, so the result is the right side expression, `new C()`.
